File: samples/simple-android/src/main/java/com/fortysevendeg/mvessel/test/db/ContactsOpenHelper.java
package com.fortysevendeg.mvessel.test.db;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import com.fortysevendeg.mvessel.Database;

public class ContactsOpenHelper extends SQLiteOpenHelper {

    public static final String TABLE = "contacts";
    public static final String C_ID = "_id";
    public static final String C_NAME = "name";
    public static final String C_AGE = "age";

    private Database database;

    public ContactsOpenHelper(Context context) {
        super(context, "contacts.db", null, 1);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        db.beginTransaction();
        db.execSQL(String.format("CREATE TABLE IF NOT EXISTS %s(%s INTEGER PRIMARY KEY AUTOINCREMENT, %s TEXT, %s INTEGER)", TABLE, C_ID, C_NAME, C_AGE));
        for (int i = 1 ; i <= 100 ; i++) {
            db.execSQL(String.format("INSERT INTO %s(%s, %s) VALUES('%s', %d)", TABLE, C_NAME, C_AGE, "Contact " + i, i + 20));
        }
        db.setTransactionSuccessful();
        db.endTransaction();
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

    }

    public Database open() {
        SQLiteDatabase database = getReadableDatabase();
        this.database = new Database(database.getPath(), 50, 0, 0);
        return this.database;
    }

    @Override
    public synchronized void close() {
        try {
            database.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
        super.close();
    }

    public Database getDatabase() {
        return database;
    }

}

File: samples/scala-android/src/main/scala/com/fortysevendeg/mvessel/sample/scala/db/ContactsOpenHelper.scala
package com.fortysevendeg.mvessel.sample.scala.db

import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import com.fortysevendeg.mvessel._
import com.fortysevendeg.mvessel.Database
import com.fortysevendeg.mvessel.api.impl.AndroidCursor
import com.fortysevendeg.mvessel.api.impl.AndroidDatabaseFactory
import ContactsOpenHelper._

import scala.util.{Failure, Try}

class ContactsOpenHelper(context: Context)
  extends SQLiteOpenHelper(context, database, javaNull, 1) {

  val database: Database[AndroidCursor] = {
    val database: SQLiteDatabase = getReadableDatabase
    new Database[AndroidCursor](
      databaseFactory = new AndroidDatabaseFactory,
      name = database.getPath,
      timeout = 50,
      retry = 0,
      flags = 0)
  }

  def onCreate(db: SQLiteDatabase) {
    db.beginTransaction()

    db.execSQL(
      s"""CREATE TABLE IF NOT EXISTS ${ContactsOpenHelper.TABLE} (
         |${ContactsOpenHelper.C_ID} INTEGER PRIMARY KEY AUTOINCREMENT,
         |${ContactsOpenHelper.C_NAME} TEXT,
         |${ContactsOpenHelper.C_AGE} INTEGER)""".stripMargin)

    1 to 100 foreach { i =>
      db.execSQL(
        s"""INSERT INTO ${ContactsOpenHelper.TABLE}(
           |${ContactsOpenHelper.C_NAME},
           |${ContactsOpenHelper.C_AGE}
           |) VALUES('Contact $i', ${i + 20})""".stripMargin)
    }

    db.setTransactionSuccessful()
    db.endTransaction()
  }

  def onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) = {}

  override def close() {
    Try(database.close()) match {
      case Failure(e) => e.printStackTrace()
      case _ =>
    }
    super.close()
  }

}

object ContactsOpenHelper {
  val database = "contacts.db"
  val TABLE: String = "contacts"
  val C_ID: String = "_id"
  val C_NAME: String = "name"
  val C_AGE: String = "age"
}